File: app/backend/schemas/emotion.py
# app/schemas/emotion.py
from datetime import datetime
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class EmotionSessionCreate(BaseModel):
    user_id: Optional[UUID] = None
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    emotion_label: Optional[str] = None
    topic: Optional[str] = None
    trigger_summary: Optional[str] = None
    insight_summary: Optional[str] = None


class EmotionSessionRead(BaseModel):
    session_id: UUID
    user_id: UUID
    started_at: datetime
    ended_at: Optional[datetime]
    emotion_label: Optional[str]
    topic: Optional[str]
    trigger_summary: Optional[str]
    insight_summary: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class EmotionStepCreate(BaseModel):
    session_id: UUID
    step_order: int
    step_type: str
    user_input: str
    gpt_response: str
    created_at: Optional[datetime] = None
    insight_tag: Optional[str] = None


class EmotionStepRead(BaseModel):
    step_id: UUID
    session_id: UUID
    step_order: int
    step_type: str
    user_input: str
    gpt_response: str
    created_at: datetime
    insight_tag: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class EmotionStepGenerateInput(BaseModel):
    session_id: Optional[UUID] = None      # ?ЖьЬ╝ый??РыПЩ ?ЭьД▒
    user_id: Optional[UUID] = None         # ???╕ьЕШ???ДьИШ
    step_type: str
    user_input: str
    temperature: Optional[float] = 0.72
    max_completion_tokens: Optional[int] = 500
    insight_tag: Optional[str] = None
    system_prompt: Optional[str] = None


class EmotionOpenRequest(BaseModel):
    type: Literal["open"] = "open"
    access_token: Optional[str] = None


class EmotionMessageRequest(BaseModel):
    type: Literal["message"] = "message"
    text: str


class EmotionCloseRequest(BaseModel):
    # Legacy/session-finalize payload. Do not use for yes/no close intent.
    type: Literal["close"] = "close"
    emotion_label: Optional[str] = None
    topic: Optional[str] = None
    trigger_summary: Optional[str] = None
    insight_summary: Optional[str] = None


class ConfirmCloseRequest(BaseModel):
    # Explicit user confirmation for the server's close suggestion.
    type: Literal["confirm_close"] = "confirm_close"


class CancelCloseRequest(BaseModel):
    # User declines the current close suggestion and keeps the session open.
    type: Literal["cancel_close"] = "cancel_close"


class TaskRecommendRequest(BaseModel):
    type: Literal["task_recommend"] = "task_recommend"
    max_items: Optional[int] = 5


class EmotionOpenResponse(BaseModel):
    type: Literal["open_ok"] = "open_ok"
    session_id: UUID
    turns: int


class EmotionMessageResponse(BaseModel):
    type: str
    delta: Optional[str] = None
    message: Optional[str] = None


class EmotionCloseResponse(BaseModel):
    # Final close acknowledgement after the session close has been persisted.
    type: Literal["close_ok"] = "close_ok"


class SuggestCloseResponse(BaseModel):
    # Server asks the client to confirm whether the session should close now.
    type: Literal["suggest_close"] = "suggest_close"


class AnalysisCardStatusResponse(BaseModel):
    type: Literal["analysis_card_status"] = "analysis_card_status"
    session_id: UUID
    status: Literal["pending", "ready", "failed"]
    card_id: Optional[UUID] = None
    message: Optional[str] = None


class TaskRecommendResponse(BaseModel):
    type: Literal["task_recommend_ok"] = "task_recommend_ok"
    items: List[dict]

File: frontend/shared/types/emotion.ts
export type EmotionSocketOutboundType =
  | "open"
  | "message"
  | "close"
  | "confirm_close"
  | "cancel_close"
  | "task_recommend";

export type EmotionSocketInboundType =
  | "open_ok"
  | "message_start"
  | "message_delta"
  | "message_end"
  | "message"
  | "suggest_close"
  | "close_ok"
  | "cancel_close_ok"
  | "analysis_card_status"
  | "analysis_card_ready"
  | "analysis_card_failed"
  | "task_recommend_ok"
  | "limit"
  | "error"
  | "ping"
  | "pong";

export interface EmotionSocketBaseMessage {
  type: string;
}

export interface EmotionSocketOpenMessage extends EmotionSocketBaseMessage {
  type: "open";
  access_token?: string;
}

export interface EmotionSocketTextMessage extends EmotionSocketBaseMessage {
  type: "message";
  text: string;
}

export interface EmotionSocketCloseMessage extends EmotionSocketBaseMessage {
  type: "close";
  emotion_label?: string;
  topic?: string;
  trigger_summary?: string;
  insight_summary?: string;
}

export interface EmotionSocketConfirmCloseMessage
  extends EmotionSocketBaseMessage {
  type: "confirm_close";
}

export interface EmotionSocketCancelCloseMessage
  extends EmotionSocketBaseMessage {
  type: "cancel_close";
}

export interface EmotionSocketTaskRecommendMessage
  extends EmotionSocketBaseMessage {
  type: "task_recommend";
  max_items?: number;
}

export type EmotionSocketOutboundMessage =
  | EmotionSocketOpenMessage
  | EmotionSocketTextMessage
  | EmotionSocketCloseMessage
  | EmotionSocketConfirmCloseMessage
  | EmotionSocketCancelCloseMessage
  | EmotionSocketTaskRecommendMessage;

export interface EmotionSocketInboundMessage extends EmotionSocketBaseMessage {
  type: EmotionSocketInboundType;
  session_id?: string;
  delta?: string;
  message?: string;
  status?: "pending" | "ready" | "failed";
  items?: Array<Record<string, unknown>>;
  [key: string]: unknown;
}

export interface EmotionSessionRead {
  session_id: string;
  user_id: string;
  started_at: string;
  ended_at: string | null;
  emotion_label: string | null;
  topic: string | null;
  trigger_summary: string | null;
  insight_summary: string | null;
}

export interface EmotionStepRead {
  step_id: string;
  session_id: string;
  step_order: number;
  step_type: string;
  user_input: string;
  gpt_response: string;
  created_at: string;
  insight_tag: string | null;
}
